packages/react-reconciler/src/ReactFiberHydrationWarning.js

```diff
@@ -0,0 +1,854 @@
+  if (Array.isArray(children)) {
+    let ic = children.length;
+    let isAllStrings = ic > 0;
+    while (--ic > 0) {
+      if (typeof children[ic] !== 'string') {
+        isAllStrings = false;
+        break;
+      }
+    }
+    if (isAllStrings) {
+      return printCurlyValue(children);
+    } else {
+      return children
+        .map(child => printChildrenValueInner(child, withinParent))
+        .join('');
+    }
+  } else {
+    return printChildrenValueInner(children, withinParent);
+  }
+}
+
+function printChildrenHTML(
+  instanceHTML: DangerouslySetInnerHTMLString,
+): string {
+  // For simplicity, assuming for display purposes that HTML is equivalent to the printed JSX markup.
+  return clipStringWithEllipsis(instanceHTML.__html, PRINT_MAX_STRING_LENGTH);
+}
+
+function printElementOpeningTag(
+  instanceDisplayName: string,
+  instanceProps: Object | null,
+  selfClosing: boolean,
+) {
+  let ret = '<' + instanceDisplayName;
+  if (instanceProps) {
+    let i = 0;
+    for (const propName in instanceProps) {
+      if (
+        !instanceProps.hasOwnProperty(propName) ||
+        REACT_SPECIFIC_PROPS.hasOwnProperty(propName)
+      ) {
+        continue;
+      }
+      if (i >= PRINT_MAX_PROP_ITEMS) {
+        ret += ' ...';
+        break;
+      }
+      ret +=
+        ' ' +
+        propName +
+        '=' +
+        printCurlyOrQuotedStringValue(instanceProps[propName]);
+      ++i;
+    }
+  }
+  return ret + (selfClosing ? ' />' : '>');
+}
+
+function printElementClosingTag(instanceDisplayName: string): string {
+  return '</' + instanceDisplayName + '>';
+}
+
+function isDangerouslySetInnerHTMLString(value: mixed): boolean %checks {
+  return (
+    typeof value === 'object' &&
+    value !== null &&
+    value.__html != null &&
+    // We need to ensure `__html` actually contains a string, as some tests put `false` in there.
+    typeof value.__html === 'string'
+  );
+}
+
+function printElementChildren(
+  instanceProps: Object | null,
+  instanceTextOrHTML: string | DangerouslySetInnerHTMLString | null,
+) {
+  if (instanceProps && instanceProps.children != null) {
+    return printChildrenValue(instanceProps.children, true);
+  } else if (
+    instanceProps &&
+    isDangerouslySetInnerHTMLString(instanceProps.dangerouslySetInnerHTML)
+  ) {
+    return printChildrenHTML(instanceProps.dangerouslySetInnerHTML);
+  } else if (typeof instanceTextOrHTML === 'string') {
+    return printChildrenValue(instanceTextOrHTML, true);
+  } else if (isDangerouslySetInnerHTMLString(instanceTextOrHTML)) {
+    return printChildrenHTML(instanceTextOrHTML);
+  } else if (instanceTextOrHTML != null) {
+    return printCurlyValue(instanceTextOrHTML);
+  } else {
+    return '';
+  }
+}
+
+function printElementOrText(
+  instanceDisplayName: string | null,
+  instanceProps: Object | null,
+  instanceTextOrHTML: string | DangerouslySetInnerHTMLString | null,
+  openingTagOnly?: boolean,
+  withinParent?: boolean,
+): string {
```

Comment:
This instance printing code happened to be more complex than I expected, but this all for the sake of beautiful and natural-looking diffs.